### DISCUSSION TITLE
Welcome to Ubiety.Stringprep.Core Discussions!

### DISCUSSION THREAD

**Discussion by coder2000** [ubiety/Ubiety.Stringprep.Core#3] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by jensfiederer** [comment]
Loaded your library, and I am trying to use it.
One thing that would be extremely helpful would be a simple example, sort of a hello world that shows how to use it.

At the moment I am guessing that you would get some things done by doing a preparationProcess.Run("input string"), but I am trying to puzzle out how to make an actual preparationProcess.

I am guessing:
var processBuilder = PreparationProcess.Build();
// a bunch of "With" steps to describe the profile you are implementing
// I am just listing one because it is available with a default :-)
processBuilder = processBuilder.WithNormalizationStep();
// a bunch more of these WITH things, presumably in the order they are specified in the RFC?  Or do you arrange them internally?
process = processBuilder.Compile();
try {
    var output = process.Run("input string");
    // yay, got it, do something with it...
} catch (Exception oneOfYourNiceExceptions) {
    // handle exception
}

**Comment by jensfiederer** [comment]
FWIW, to do a SaslPrep profile (used for PDF encryption password handling) I did:

            var build = PreparationProcess.Build();
            build = build.WithMappingStep(MappingTable.Create(Tables.C12.ToArray(), 0x20));
            build = build.WithMappingStep(MappingTable.Create(Tables.B1));
            build = build.WithNormalizationStep();
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C12));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C21));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C22));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C3));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C4));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C5));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C6));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C7));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C8));
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Prohibited.C9));
            build = build.WithBidirectionalStep();
            build = build.WithProhibitedValueStep(ValueRangeTable.Create(Unassigned.UnassignedCodePoints));
            _proc = build.Compile();

and it worked, so thank you (not 100% sure about prohibiting the Unassigned, but seems to make sense)

**Reply by coder2000** [reply]
This is how I use it to build a Nameprep profile in Ubiety.Xmpp.Core:
```cs
            return PreparationProcess.Build()
                .WithMappingStep(MappingTable.Build()
                    .WithMappingTable(Mapping.B_1)
                    .WithMappingTable(Mapping.B_2)
                    .Compile())
                .WithNormalizationStep()
                .WithProhibitedValueStep(ValueRangeTable.Create(
                    Prohibited.C_1_1,
                    Prohibited.C_2_2,
                    Prohibited.C_3,
                    Prohibited.C_4,
                    Prohibited.C_5,
                    Prohibited.C_6,
                    Prohibited.C_7,
                    Prohibited.C_8,
                    Prohibited.C_9))
                .WithBidirectionalStep()
                .WithProhibitedValueStep(ValueRangeTable.Create(
                    Unassigned.A_1))
                .Compile();
```

Sorry about not having an example as part of the code.

**Reply by jensfiederer** [reply]
Oh, it was VERY helpful to have your code, I got working code, thank you again!

I do like your flow, mine kind of stutters a little with all the "build = build....." items!
